File: _prod_validate.js
require('dotenv').config({override:true,quiet:true});
const http=require('http');const fs=require('fs');const path=require('path');

let cookie='';
function req(method,path2,body){
  return new Promise(resolve=>{
    const u=new URL('http://localhost:3099'+path2);
    const opts={hostname:u.hostname,port:u.port,path:u.pathname+u.search,method};
    const hdrs={'Content-Type':'application/json'};
    if(body){hdrs['Content-Length']=Buffer.from(JSON.stringify(body)).length}
    if(cookie)hdrs.Cookie=cookie;
    const q=http.request({...opts,headers:hdrs},resp=>{
      const sc=resp.headers['set-cookie'];
      const scStr=(Array.isArray(sc)?sc[0]:sc)||'';
      const m=scStr.match(/hisecure\.sid=([^;]+)/);
      if(m)cookie='hisecure.sid='+m[1];
      const d=[];resp.on('data',c=>d.push(c));
      resp.on('end',()=>{try{resolve({status:resp.statusCode,body:JSON.parse(Buffer.concat(d).toString())})}catch(e){resolve({status:resp.statusCode,body:Buffer.concat(d).toString()})}});
    });
    q.on('error',()=>resolve({status:0,body:{error:'connection'}}));
    if(body)q.write(JSON.stringify(body));q.end();
  });
}

(async()=>{
  let lr=await req('POST','/api/auth/login',{username:'admin',password:'admin@123'});
  if(lr.status!==200){console.log('LOGIN FAIL');process.exit(1)}
  console.log('LOGIN:',lr.status,'cookie set:',!!cookie,'user='+lr.body.user.username,'id='+lr.body.user.id);
  let pass=0,fail=0;
  let out={suite:'all',login:lr.status,results:{},passCount:0,failCount:0,total:0,timestamp:new Date().toISOString()};
  function chk(name,status,detail){
    if(status===200)pass++;else fail++;
    out.results[name]={status,pass:status===200,detail:(detail||'')};
    out.passCount=pass;out.failCount=fail;out.total=pass+fail;
    out.verdict=fail===0?'PASS':'FAIL';
    console.log(' '+name+': '+(status===200?'PASS':'FAIL')+' ['+status+'] '+(detail||''));
  }

  console.log('\n=== RBAC ===');
  let r;
  r=await req('GET','/api/users?limit=1');chk('users',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/technicians');chk('technicians',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/complaints?limit=1');chk('complaints',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/amc/contracts?limit=1');chk('amc',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/repairs?limit=1');chk('repairs',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/tickets?limit=1');chk('tickets',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/settings');chk('settings',r.status,r.body.error||r.body.message||'ok');
  r=await req('GET','/api/dashboard');chk('dashboard',r.status,r.body.error||'ok');
  r=await req('GET','/api/reports/stats');chk('reports',r.status,r.body.error||'ok');
  r=await req('GET','/api/products?limit=1');chk('products',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/customers?limit=1');chk('customers',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/payments?limit=1');chk('payments',r.status,r.body.error||r.body.message||'ok');
  r=await req('GET','/api/parts?limit=1');chk('parts',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/suppliers?limit=1');chk('suppliers',r.status,r.body.error||r.body.message||'ok');
  r=await req('GET','/api/stores?limit=1');chk('stores',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/invoices?limit=1');chk('invoices',r.status,r.body.error||'rows='+((r.body.data||[]).length));
  r=await req('GET','/api/accounting?limit=1');chk('accounting',r.status,r.body.error||r.body.message||'ok');

  let custId=null;
  r=await req('GET','/api/customers?limit=1');
  if(r.status===200&&r.body.data&&r.body.data[0])custId=r.body.data[0].customer_id;
  console.log('\nCustomer:',custId);

  if(custId){
    console.log('\n=== Workflow 2: Complaint -> Ticket ===');
    let c=await req('POST','/api/complaints',{customer_id:custId,subject:'wb-Complaint',priority:'medium',category:'service'});
    let compId=(c.body&&c.body.data&&c.body.data.complaint_id)||null;
    chk('wf2_create_complaint',c.status,compId?'compId='+compId:(c.body.error||'fail'));

    if(compId){
      let r2=await req('GET','/api/complaints/'+compId);chk('wf2_read',r2.status,(r2.body.data&&r2.body.data.complaint_id?'ok':r2.body.error));
      r2=await req('PUT','/api/complaints/'+compId+'/status',{status:'under_review'});chk('wf2_review',r2.status,(r2.body.error||'ok'));
      r2=await req('PUT','/api/complaints/'+compId+'/status',{status:'resolved',resolution:'wb'});chk('wf2_resolved',r2.status,(r2.body.error||'ok'));

      let t=await req('POST','/api/tickets',{customer_id:custId,subject:'wb-Ticket',priority:'medium',ticket_type:'service',complaint_id:compId,description:'wb'});
      let tId=(t.body&&t.body.data&&t.body.data.ticket_id)||null;
      chk('wf2_ticket',t.status,tId?'tId='+tId:(t.body.error||'fail'));

      if(tId){
        let r3=await req('GET','/api/tickets/'+tId);chk('wf2_ticket_read',r3.status,(r3.body.data&&r3.body.data.ticket_id?'ok':r3.body.error));
        r3=await req('PUT','/api/tickets/'+tId,{status:'assigned'});chk('wf2_assigned',r3.status,(r3.body.error||'ok'));
        r3=await req('PUT','/api/tickets/'+tId,{status:'in_progress'});chk('wf2_progress',r3.status,(r3.body.error||'ok'));
        r3=await req('PUT','/api/tickets/'+tId,{status:'closed'});chk('wf2_closed',r3.status,(r3.body.error||'ok'));
        r3=await req('GET','/api/tickets?limit=5&status=closed');chk('wf2_closed_filter',r3.status,(r3.body.error||'n='+((r3.body.data||[]).length)));
      }
    }

    console.log('\n=== Workflow 3: AMC ===');
    let a=await req('POST','/api/amc/contracts',{customer_id:custId,contract_type:'annual',start_date:'2026-01-01',end_date:'2027-01-01',terms:'wb'});
    let amcId=(a.body&&a.body.data&&a.body.data.amc_id)||null;
    chk('wf3_amc',a.status,amcId?'amcId='+amcId:(a.body.error||'fail'));

    if(amcId){
      let r4=await req('GET','/api/amc/contracts/'+amcId);chk('wf3_amc_read',r4.status,(r4.body.data&&r4.body.data.amc_id?'ok':r4.body.error));
      r4=await req('POST','/api/amc/contracts/'+amcId+'/activate');chk('wf3_activate',r4.status,(r4.body.error||'ok'));

      let aset=await req('POST','/api/amc/assets',{amc_id:amcId,asset_type:'equipment',serial_number:'wb-'+Date.now(),is_active:true});
      let asetId=(aset.body&&aset.body.data&&aset.body.data.asset_id)||null;
      chk('wf3_asset',aset.status,asetId?'astId='+asetId:(aset.body.error||'fail'));

      if(asetId){
        let r5=await req('GET','/api/amc/assets/'+asetId);chk('wf3_asset_read',r5.status,(r5.body.data&&r5.body.data.asset_id?'ok':r5.body.error));
        r5=await req('GET','/api/amc/assets?amc_id='+amcId);chk('wf3_asset_list',r5.status,(r5.body.error||'n='+((r5.body.data||[]).length)));
      }
      r4=await req('GET','/api/amc/stats');chk('wf3_stats',r4.status,(r4.body.error||'ok'));
      r4=await req('PUT','/api/amc/contracts/'+amcId,{terms:'wb-up'});chk('wf3_update',r4.status,(r4.body.error||'ok'));
      r4=await req('GET','/api/amc/contracts?customer_id='+custId);chk('wf3_filter',r4.status,(r4.body.error||'ok'));
    }
  }

  fs.writeFileSync('C:/Users/Admin/Desktop/Calude Test/erp-app/_results_all.json',JSON.stringify(out,null,2));
  console.log('\n=== FINAL ===');
  console.log('Total:',pass+'/'+out.total,' PASS,',fail,'FAIL');
  console.log('Verdict:',out.verdict);
})();

File: _prod_validate.py
#!/usr/bin/env python3
"""HiSecure ERP -- Production Validation (fresh TCP per request)"""
import http.client, json, time, re, os, sys
sys.stdout.reconfigure(encoding='utf-8')

BASE = 'localhost:3099'
RESULT = 'C:/Users/Admin/Desktop/Calude Test/erp-app/_results_prod.json'
COOKIE_FILE = 'C:/Users/Admin/Desktop/Calude Test/erp-app/.cookie.txt'

cookie_str = ''
pass_count = 0
fail_count = 0
results = {}

def save_cookie(headers):
    """Extract session cookie from response (handles set-cookie string or array)."""
    global cookie_str
    raw = headers.get('set-cookie') or headers.get('Set-Cookie') or ''
    if isinstance(raw, list):
        raw = raw[0] if raw else ''
    m = re.search(r'hisecure\.sid=([^;]+)', raw)
    if m:
        cookie_str = 'hisecure.sid=' + m[1]
        with open(COOKIE_FILE, 'w') as f:
            f.write(cookie_str)

def load_cookie():
    global cookie_str
    if os.path.exists(COOKIE_FILE):
        with open(COOKIE_FILE) as f:
            cookie_str = f.read().strip()

def req(method, path, body=None):
    """Fresh HTTP connection per call -- avoids keep-alive pool bug."""
    global cookie_str
    conn = http.client.HTTPConnection('localhost', 3099, timeout=8)
    hdrs = {
        'Content-Type': 'application/json',
        'Connection': 'close',           # Force new TCP per request
    }
    if cookie_str:
        hdrs['Cookie'] = cookie_str
    data = json.dumps(body).encode() if body else None
    try:
        conn.request(method, path, body=data, headers=hdrs)
        resp = conn.getresponse()
        body_text = resp.read().decode()
        save_cookie(resp.headers)
        try:
            body_json = json.loads(body_text)
        except json.JSONDecodeError:
            body_json = body_text
        return resp.status, body_json
    except Exception as e:
        return 0, {'error': str(e)}
    finally:
        conn.close()

def chk(name, code, detail=''):
    global pass_count, fail_count, results
    ok = (code == 200)
    if ok:
        pass_count += 1
    else:
        fail_count += 1
    results[name] = {'status': code, 'pass': ok, 'detail': detail}
    print(f'  {name}: {"PASS" if ok else "FAIL"} [HTTP {code}] {detail}')

# ---- MAIN ----
load_cookie()
print('=== Login & Session ===')
code, body = req('POST', '/api/auth/login', {'username': 'admin', 'password': 'admin@123'})
print(f'  POST /api/auth/login -> {code} user={body.get("user", {}).get("username", "?")}')
if code != 200:
    print('FATAL: Login failed')
    exit(1)
print(f'  COOKIE: {cookie_str[:50]}...')

print('\n=== RBAC: Module Access ===')
chk('users', *(req('GET', '/api/users?limit=1')))
chk('technicians', *(req('GET', '/api/technicians')))
chk('complaints', *(req('GET', '/api/complaints?limit=1')))
chk('amc_contracts', *(req('GET', '/api/amc/contracts?limit=1')))
chk('repairs', *(req('GET', '/api/repairs?limit=1')))
chk('tickets', *(req('GET', '/api/tickets?limit=1')))
chk('settings', *(req('GET', '/api/settings')))
chk('dashboard', *(req('GET', '/api/dashboard')))
chk('reports', *(req('GET', '/api/reports/stats')))
chk('products', *(req('GET', '/api/products?limit=1')))
chk('customers', *(req('GET', '/api/customers?limit=1')))
chk('payments', *(req('GET', '/api/payments?limit=1')))
chk('parts', *(req('GET', '/api/parts?limit=1')))
chk('suppliers', *(req('GET', '/api/suppliers?limit=1')))
chk('stores', *(req('GET', '/api/stores?limit=1')))
chk('invoices', *(req('GET', '/api/invoices?limit=1')))
chk('accounting', *(req('GET', '/api/accounting?limit=1')))

print('\n=== Workflow 1: Quote -> Invoice -> DC ===')
cust_code, cust_body = req('GET', '/api/customers?limit=1')
cust_id = None
if cust_code == 200 and cust_body.get('data') and cust_body['data']:
    cust_id = cust_body['data'][0].get('customer_id')
    print(f'  Customer: {cust_id}')

if cust_id:
    parts_code, parts_body = req('GET', '/api/parts?limit=1')
    part_id = None
    if parts_code == 200 and parts_body.get('data') and parts_body['data']:
        part_id = parts_body['data'][0].get('part_id')

    q_code, q_body = req('POST', '/api/quotations', {
        'customer_id': cust_id,
        'items': [{'part_id': part_id, 'quantity': 1, 'unit_price': 100, 'discount_percent': 0, 'tax_rate': 18}],
        'terms': 'py-wf1', 'notes': 'prod-test'
    })
    quote_id = None
    if q_code == 200 and q_body.get('data'):
        quote_id = q_body['data'].get('quote_id')
        chk('wf1_quote', q_code, f'quote_id={quote_id}')
    else:
        chk('wf1_quote', q_code, q_body.get('error', 'fail'))

    if quote_id:
        req('PUT', f'/api/quotations/{quote_id}/status', {'status': 'sent'})
        req('PUT', f'/api/quotations/{quote_id}/status', {'status': 'accepted'})
        cv_code, cv_body = req('POST', f'/api/quotations/{quote_id}/convert')
        invoice_id = None
        if cv_code == 200 and cv_body.get('data'):
            invoice_id = cv_body['data'].get('invoice_id')
            chk('wf1_convert', cv_code, f'inv_id={invoice_id}')
        else:
            chk('wf1_convert', cv_code, cv_body.get('error', 'fail'))
        if invoice_id:
            req('POST', f'/api/invoices/{invoice_id}/issue')

        dc_code, dc_body = req('POST', '/api/delivery-challans', {
            'from_location_id': 1, 'to_location_id': 1, 'challan_date': '2026-06-05',
            'items': [{'part_id': part_id, 'quantity': 1}], 'purposes': ['sale']
        })
        dc_id = None
        if dc_code == 200 and dc_body.get('data'):
            dc_id = dc_body['data'].get('delivery_challan_id')
            chk('wf1_dc', dc_code, f'dc_id={dc_id}')
        else:
            chk('wf1_dc', dc_code, dc_body.get('error', 'fail'))
        if dc_id:
            req('PUT', f'/api/delivery-challans/{dc_id}/status', {'status': 'dispatched'})
else:
    print('  SKIP WF1: no customer')

print('\n=== Workflow 2: Complaint -> Ticket ===')
if cust_id:
    c_code, c_body = req('POST', '/api/complaints', {
        'customer_id': cust_id, 'subject': 'py-wf2', 'priority': 'medium', 'category': 'service'
    })
    comp_id = None
    if c_code == 200 and c_body.get('data'):
        comp_id = c_body['data'].get('complaint_id')
        chk('wf2_complaint', c_code, f'comp_id={comp_id}')
    else:
        chk('wf2_complaint', c_code, c_body.get('error', 'fail'))

    if comp_id:
        chk('wf2_comp_read', *(req('GET', f'/api/complaints/{comp_id}')))
        chk('wf2_comp_review', *(req('PUT', f'/api/complaints/{comp_id}/status', {'status': 'under_review'})))
        chk('wf2_comp_resolved', *(req('PUT', f'/api/complaints/{comp_id}/status', {'status': 'resolved', 'resolution': 'py'})))

        t_code, t_body = req('POST', '/api/tickets', {
            'customer_id': cust_id, 'subject': 'py-wf2-tk', 'priority': 'medium',
            'ticket_type': 'service', 'complaint_id': comp_id, 'description': 'py'
        })
        ticket_id = None
        if t_code == 200 and t_body.get('data'):
            ticket_id = t_body['data'].get('ticket_id')
            chk('wf2_ticket', t_code, f'tk_id={ticket_id}')
        else:
            chk('wf2_ticket', t_code, t_body.get('error', 'fail'))

        if ticket_id:
            chk('wf2_tk_read', *(req('GET', f'/api/tickets/{ticket_id}')))
            chk('wf2_tk_assigned', *(req('PUT', f'/api/tickets/{ticket_id}', {'status': 'assigned'})))
            chk('wf2_tk_progress', *(req('PUT', f'/api/tickets/{ticket_id}', {'status': 'in_progress'})))
            chk('wf2_tk_closed', *(req('PUT', f'/api/tickets/{ticket_id}', {'status': 'closed'})))
else:
    print('  SKIP WF2: no customer')

print('\n=== Workflow 3: AMC ===')
if cust_id:
    a_code, a_body = req('POST', '/api/amc/contracts', {
        'customer_id': cust_id, 'contract_type': 'annual',
        'start_date': '2026-01-01', 'end_date': '2027-01-01', 'terms': 'py-wf3'
    })
    amc_id = None
    if a_code == 200 and a_body.get('data'):
        amc_id = a_body['data'].get('amc_id')
        chk('wf3_amc', a_code, f'amc_id={amc_id}')
    else:
        chk('wf3_amc', a_code, a_body.get('error', 'fail'))

    if amc_id:
        chk('wf3_amc_read', *(req('GET', f'/api/amc/contracts/{amc_id}')))
        chk('wf3_activate', *(req('POST', f'/api/amc/contracts/{amc_id}/activate')))

        as_code, as_body = req('POST', '/api/amc/assets', {
            'amc_id': amc_id, 'asset_type': 'equipment',
            'serial_number': f'py-{int(time.time())}', 'is_active': True
        })
        asset_id = None
        if as_code == 200 and as_body.get('data'):
            asset_id = as_body['data'].get('asset_id')
            chk('wf3_asset', as_code, f'asset_id={asset_id}')
        else:
            chk('wf3_asset', as_code, as_body.get('error', 'fail'))

        if asset_id:
            chk('wf3_asset_read', *(req('GET', f'/api/amc/assets/{asset_id}')))
            chk('wf3_asset_list', *(req('GET', f'/api/amc/assets?amc_id={amc_id}')))

        chk('wf3_stats', *(req('GET', '/api/amc/stats')))
        chk('wf3_update', *(req('PUT', f'/api/amc/contracts/{amc_id}', {'terms': 'py-up'})))
        chk('wf3_filter', *(req('GET', f'/api/amc/contracts?customer_id={cust_id}')))
else:
    print('  SKIP WF3: no customer')

# ---- SUMMARY ----
total = pass_count + fail_count
verdict = 'PASS' if fail_count == 0 else 'FAIL'
report = {
    'suite': 'ProductionValidation',
    'timestamp': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
    'server': 'http://localhost:3099',
    'pass': pass_count,
    'fail': fail_count,
    'total': total,
    'verdict': verdict,
    'results': results
}
with open(RESULT, 'w') as f:
    json.dump(report, f, indent=2)

print(f'\n{"="*50}')
print(f'RESULTS: {pass_count}/{total} PASS, {fail_count} FAIL')
print(f'Verdict: {verdict}')
print(f'File: {RESULT}')
print(f'{"="*50}')
